### PR TITLE
CB-2534 Support sql server and oracle as internal db

### PR DESCRIPTION
**Pull request by oksana-cherniavskaia**
CB-2534 Support sql server and oracle as internal db

### CODE REVIEW

File: plugins/org.jkiss.dbeaver.ext.mssql/src/org/jkiss/dbeaver/ext/mssql/model/SQLServerDialect.java [hunk 1]
```diff
@@ -42,10 +43,10 @@
 
 import java.util.*;
 
-public class SQLServerDialect extends JDBCSQLDialect implements TPRuleProvider {
+public class SQLServerDialect extends JDBCSQLDialect implements TPRuleProvider, SQLDialectDDLExtension {
 
     private static final String[][] TSQL_BEGIN_END_BLOCK = new String[][]{
-        {SQLConstants.BLOCK_BEGIN, SQLConstants.BLOCK_END}
+            {SQLConstants.BLOCK_BEGIN, SQLConstants.BLOCK_END}
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ext.mssql/src/org/jkiss/dbeaver/ext/mssql/model/SQLServerDialect.java]: Please fix your formatting configuration. The `continuation indent` should be `4`, you have `8`.

File: plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/translate/SQLQueryTranslationManager.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+     * Translates script to target dialect.
+     *
+     * @param sourceDialect   the source dialect
+     * @param targetDialect   the target dialect
+     * @param preferenceStore the preference store
+     * @param script          the script
+     * @return the string
+     * @throws DBException the db exception
+     */
+    public static String translateScript(
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/translate/SQLQueryTranslationManager.java]: Please specify nullability using an appropriate annotation

File: plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/translate/SQLQueryTranslationManager.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+            @NotNull SQLDialect sourceDialect,
+            @NotNull SQLDialect targetDialect,
+            @NotNull DBPPreferenceStore preferenceStore,
+            @NotNull String script) throws DBException {
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/translate/SQLQueryTranslationManager.java]:
```suggestion
            @NotNull String script
        ) throws DBException {
```

File: plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/translate/SQLQueryTranslator.java [hunk 4]
```diff
@@ -14,47 +14,182 @@
...
-                context.translateCommand(element));
+public class SQLQueryTranslator implements SQLTranslator {
+
+    private SQLTranslateContext sqlTranslateContext;
+
+    public SQLQueryTranslator(SQLTranslateContext sqlTranslateContext) {
+        this.sqlTranslateContext = sqlTranslateContext;
+    }
+
+    public List<? extends SQLScriptElement> translate(@NotNull SQLScriptElement element) throws DBException {
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/translate/SQLQueryTranslator.java:101]: Please specify return type nullability using an appropriate annotation

File: plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/translate/SQLQueryTranslator.java [hunk 5]
```diff
@@ -14,47 +14,182 @@
...
-        List<SQLScriptElement> sqlScriptElements = SQLScriptParser.parseScript(sourceDialect, preferenceStore, script);
-        List<SQLScriptElement> result = new ArrayList<>();
-        for (SQLScriptElement element : sqlScriptElements) {
-            result.addAll(
-                context.translateCommand(element));
+public class SQLQueryTranslator implements SQLTranslator {
+
+    private SQLTranslateContext sqlTranslateContext;
+
+    public SQLQueryTranslator(SQLTranslateContext sqlTranslateContext) {
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/translate/SQLQueryTranslator.java]: Please specify nullability using an appropriate annotation

File: plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/translate/SQLQueryTranslator.java [hunk 6]
```diff
@@ -14,47 +14,182 @@
...
-        String scriptDelimiter = targetDialect.getScriptDelimiters()[0];
 
-        StringBuilder sql = new StringBuilder();
-        for (SQLScriptElement element : result) {
-            sql.append(element.getText());
-            sql.append(scriptDelimiter).append("\n");
+        return Collections.singletonList(element);
+    }
+
+    private List<? extends SQLScriptElement> translateQuery(@NotNull SQLQuery query) {
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/translate/SQLQueryTranslator.java:111]: Please specify return type nullability using an appropriate annotation

File: plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/translate/SQLQueryTranslator.java [hunk 7]
```diff
@@ -14,47 +14,182 @@
...
 
+    /**
+     * Translates statement to target dialect.
+     *
+     * @param query     the query
+     * @param statement the statement
+     * @return the list
+     */
+    protected List<? extends SQLScriptElement> translateStatement(@NotNull SQLQuery query,
+                                                                  @NotNull Statement statement) {
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/translate/SQLQueryTranslator.java]: Please specify nullability using an appropriate annotation

```suggestion
    protected List<? extends SQLScriptElement> translateStatement(
        @NotNull SQLQuery query,
        @NotNull Statement statement
    ) {
```

File: plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/translate/SQLQueryTranslator.java [hunk 8]
```diff
@@ -14,47 +14,182 @@
...
+                        if (extendedDialect != null && extendedDialect.timestampAsDatetime()) {
+                            newDataType = "datetime";
+                        }
+                        break;
+                    case "BIGINT":
+                        if (extendedDialect != null) {
+                            newDataType = extendedDialect.getLargeNumericType();
+                        }
+                        break;
+                }
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/translate/SQLQueryTranslator.java:173]: Please add an empty `default` branch to make it clear that no action should be performed.

File: plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/translate/SQLQueryTranslator.java [hunk 9]
```diff
@@ -14,47 +14,182 @@
...
+                                        extraQueries = new ArrayList<>();
+                                    }
+                                    extraQueries.add(new SQLQuery(null, createSeqQuery));
+                                } else if (extendedDialect != null) {
+                                    int indexOf = cd.getColumnSpecs().indexOf(cSpec);
+                                    defChanged = true;
+                                    cd.getColumnSpecs().set(indexOf, extendedDialect.getAutoIncrementKeyword());
+                                }
+                                break;
+                        }
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/translate/SQLQueryTranslator.java:208]: Same as previous comment

File: plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/translate/SQLQueryTranslator.java [hunk 10]
```diff
@@ -14,47 +14,182 @@
...
+        }
+        return extraQueries;
+    }
+
+    /**
+     * Gets sql translate context.
+     *
+     * @return the sql translate context
+     */
+    public SQLTranslateContext getSqlTranslateContext() {
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/translate/SQLQueryTranslator.java:237]: Please specify nullability using an appropriate annotation

File: plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/translate/SQLQueryTranslator.java [hunk 11]
```diff
@@ -14,47 +14,182 @@
...
+        return sqlTranslateContext;
+    }
+
+    /**
+     * Sets sql translate context.
+     *
+     * @param sqlTranslateContext the sql translate context
+     */
+    public void setSqlTranslateContext(SQLTranslateContext sqlTranslateContext) {
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/translate/SQLQueryTranslator.java]: Please specify nullability using an appropriate annotation

File: plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/translate/SQLTranslator.java [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+public interface SQLTranslator {
+
+    /**
+     * Translate script element to target dialect.
+     *
+     * @param element the element
+     * @return the list
+     * @throws DBException the db exception
+     */
+    List<? extends SQLScriptElement> translate(@NotNull SQLScriptElement element) throws DBException;
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/translate/SQLTranslator.java:38]: Please specify nullability using an appropriate annotation

File: plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/sql/SQLDialectDDLExtension.java [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jkiss.dbeaver.model.sql;
+
+/**
+ * Sql dialect ddl extension.
+ */
+public interface SQLDialectDDLExtension {
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/sql/SQLDialectDDLExtension.java]: Do we really need this extension interface?

I suggest adding these methods to the `SQLDialect` interface and removing this extension.

If you want to keep it, please make it extend `SQLDialect`.

**Review by ShadelessFox — CHANGES_REQUESTED** [review]

File: plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/sql/SQLDialectDDLExtension.java (hunk 13 above)
Comment (serge-rider) [reply plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/sql/SQLDialectDDLExtension.java]: Yes, we do. DDL extension may add tens of new function. It doesn't make sense for most dialects.
Especially for NoSQL ones.

File: plugins/org.jkiss.dbeaver.ext.mssql/src/org/jkiss/dbeaver/ext/mssql/model/SQLServerDialect.java (hunk 1 above)
Comment (serge-rider) [inline plugins/org.jkiss.dbeaver.ext.mssql/src/org/jkiss/dbeaver/ext/mssql/model/SQLServerDialect.java]: Continuous indent = 4 spaces.

File: plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/translate/SQLQueryTranslationManager.java [hunk 14]
```diff
@@ -0,0 +1,71 @@
+/*
```
Comment (serge-rider) [inline plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/translate/SQLQueryTranslationManager.java]: Why do we need a new  class jsut for one function?
Let's move it to SQLTranslatorHandler class.

Comment (oksana-cherniavskaia) [reply plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/translate/SQLQueryTranslationManager.java:1]: SQLTranslatorHandler located in org.jkiss.dbeaver.ui.app.standalone.cli and seems to be not suitable place for our translation as we do not work with cli stuff.

File: plugins/org.jkiss.dbeaver.ext.mssql/src/org/jkiss/dbeaver/ext/mssql/model/SQLServerDialect.java (hunk 1 above)
Comment (oksana-cherniavskaia) [reply plugins/org.jkiss.dbeaver.ext.mssql/src/org/jkiss/dbeaver/ext/mssql/model/SQLServerDialect.java]: fixed

File: plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/translate/SQLQueryTranslationManager.java (hunk 2 above)
Comment (oksana-cherniavskaia) [reply plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/translate/SQLQueryTranslationManager.java]: fixed

File: plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/translate/SQLQueryTranslationManager.java (hunk 3 above)
Comment (oksana-cherniavskaia) [reply plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/translate/SQLQueryTranslationManager.java]: fixed

File: plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/translate/SQLQueryTranslator.java (hunk 4 above)
Comment (oksana-cherniavskaia) [reply plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/translate/SQLQueryTranslator.java:101]: fixed

File: plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/translate/SQLQueryTranslator.java (hunk 5 above)
Comment (oksana-cherniavskaia) [reply plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/translate/SQLQueryTranslator.java]: fixed

File: plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/translate/SQLQueryTranslator.java (hunk 6 above)
Comment (oksana-cherniavskaia) [reply plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/translate/SQLQueryTranslator.java:111]: fixed

File: plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/translate/SQLQueryTranslator.java (hunk 10 above)
Comment (oksana-cherniavskaia) [reply plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/translate/SQLQueryTranslator.java:237]: fixed

File: plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/translate/SQLQueryTranslator.java (hunk 9 above)
Comment (oksana-cherniavskaia) [reply plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/translate/SQLQueryTranslator.java:208]: fixed

File: plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/translate/SQLQueryTranslator.java (hunk 7 above)
Comment (oksana-cherniavskaia) [reply plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/translate/SQLQueryTranslator.java]: fixed

File: plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/translate/SQLQueryTranslator.java (hunk 11 above)
Comment (oksana-cherniavskaia) [reply plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/translate/SQLQueryTranslator.java]: fixed

File: plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/translate/SQLTranslator.java (hunk 12 above)
Comment (oksana-cherniavskaia) [reply plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/translate/SQLTranslator.java:38]: fixed

File: plugins/org.jkiss.dbeaver.ext.mssql/src/org/jkiss/dbeaver/ext/mssql/model/SQLServerDialect.java (hunk 1 above)
Comment (oksana-cherniavskaia) [reply plugins/org.jkiss.dbeaver.ext.mssql/src/org/jkiss/dbeaver/ext/mssql/model/SQLServerDialect.java]: fixed

File: plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/translate/SQLQueryTranslationManager.java (hunk 14 above)
Comment (oksana-cherniavskaia) [reply plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/translate/SQLQueryTranslationManager.java]: Moved to SQLQueryTranslator

File: plugins/org.jkiss.dbeaver.ext.mssql/src/org/jkiss/dbeaver/ext/mssql/model/SQLServerDialect.java [hunk 15]
```diff
@@ -40,31 +41,35 @@
 import org.jkiss.utils.ArrayUtils;
 import org.jkiss.utils.CommonUtils;
 
-import java.util.*;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Locale;
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ext.mssql/src/org/jkiss/dbeaver/ext/mssql/model/SQLServerDialect.java]: We do use `*` imports for 5 and more imports from the same package.

File: plugins/org.jkiss.dbeaver.ext.mssql/src/org/jkiss/dbeaver/ext/mssql/model/SQLServerDialect.java [hunk 16]
```diff
@@ -190,7 +195,11 @@ public MultiValueInsertMode getDefaultMultiValueInsertMode() {
     }
 
     @Override
-    public String getColumnTypeModifiers(@NotNull DBPDataSource dataSource, @NotNull DBSTypedObject column, @NotNull String typeName, @NotNull DBPDataKind dataKind) {
+    public String getColumnTypeModifiers(@NotNull DBPDataSource dataSource,
+                                         @NotNull DBSTypedObject column,
+                                         @NotNull String typeName,
+                                         @NotNull DBPDataKind dataKind
+    ) {
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ext.mssql/src/org/jkiss/dbeaver/ext/mssql/model/SQLServerDialect.java]:
```suggestion
    public String getColumnTypeModifiers(
        @NotNull DBPDataSource dataSource,
        @NotNull DBSTypedObject column,
        @NotNull String typeName,
        @NotNull DBPDataKind dataKind
    ) {
```

File: plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/parser/SQLScriptParser.java [hunk 17]
```diff
@@ -806,19 +806,19 @@ public static List<SQLScriptElement> parseScript(DBPDataSource dataSource, Strin
...
-
-        SQLParserContext parserContext = new SQLParserContext(null, syntaxManager, ruleManager, sqlDocument);
-        parserContext.setPreferenceStore(preferenceStore);
+        SQLParserContext parserContext = prepareSqlParserContext(dialect, preferenceStore, sqlScriptContent);
         return SQLScriptParser.extractScriptQueries(parserContext, 0, sqlScriptContent.length(), true, false, true);
     }
 
-    public static SQLScriptElement parseQuery(SQLDialect dialect, DBPPreferenceStore preferenceStore, String sqlScriptContent, int cursorPosition) {
+    public static SQLScriptElement parseQuery(SQLDialect dialect, DBPPreferenceStore preferenceStore, String sqlScriptContent,
+                                              int cursorPosition) {
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/parser/SQLScriptParser.java]: Please add nullability annotations

```suggestion
    public static SQLScriptElement parseQuery(
        SQLDialect dialect,
        DBPPreferenceStore preferenceStore,
        String sqlScriptContent,
        int cursorPosition
    ) {
```

File: plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/parser/SQLScriptParser.java [hunk 18]
```diff
@@ -806,19 +806,19 @@ public static List<SQLScriptElement> parseScript(DBPDataSource dataSource, Strin
...
+        SQLParserContext parserContext = prepareSqlParserContext(dialect, preferenceStore, sqlScriptContent);
+        return SQLScriptParser.extractQueryAtPos(parserContext, cursorPosition);
+    }
+
+    @NotNull
+    private static SQLParserContext prepareSqlParserContext(SQLDialect dialect, DBPPreferenceStore preferenceStore,
+                                                            String sqlScriptContent) {
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/parser/SQLScriptParser.java:841]: Please add nullability annotations

```suggestion
    private static SQLParserContext prepareSqlParserContext(
        SQLDialect dialect,
        DBPPreferenceStore preferenceStore,
        String sqlScriptContent
    ) {
```

File: plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/translate/SQLQueryTranslator.java (hunk 6 above)
Comment (ShadelessFox) [reply plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/translate/SQLQueryTranslator.java:111]: There's no return type annotation

File: plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/translate/SQLQueryTranslator.java [hunk 19]
```diff
@@ -56,5 +88,159 @@ public static String translateScript(
...
+    /**
+     * Translates statement to target dialect.
+     *
+     * @param query     the query
+     * @param statement the statement
+     * @return the list
+     */
+    @NotNull
+    protected List<? extends SQLScriptElement> translateStatement(@NotNull SQLQuery query,
+                                                                  @NotNull Statement statement
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/translate/SQLQueryTranslator.java]:
```suggestion
    protected List<? extends SQLScriptElement> translateStatement(
        @NotNull SQLQuery query,
        @NotNull Statement statement
```

File: plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/sql/SQLDialectDDLExtension.java (hunk 13 above)
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/sql/SQLDialectDDLExtension.java]:
```suggestion
public interface SQLDialectDDLExtension extends SQLDialect {
```

File: plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/sql/SQLDialectDDLExtension.java [hunk 20]
```diff
@@ -0,0 +1,59 @@
...
+
+    /**
+     * Gets auto increment keyword.
+     *
+     * @return the auto increment keyword
+     */
+    String getAutoIncrementKeyword();
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/sql/SQLDialectDDLExtension.java:33]: Please add nullability annotations

**Review by ShadelessFox — CHANGES_REQUESTED** [review]

**Review by ShadelessFox — APPROVED** [review]

**Pull request merged**